### ISSUE TITLE
Investigate `mapValues` type mapping issue

### ISSUE THREAD

**Issue by nflaig** [ChainSafe/lodestar#6832]
Current implementation or rather the typing / generics of `mapValues` function to seem to be incorrect, requiring to type cast the returned value (although the implementation itself works as expected)
https://github.com/ChainSafe/lodestar/blob/c63509800c61eb44e59d41187bb272c0d1e1e462/packages/api/src/utils/client/client.ts#L124

This is the current implementation which needs to be reviewed / reworked

https://github.com/ChainSafe/lodestar/blob/8b6ecc44c1892115d0e7396b2a5f0ae8fc0d9e6f/packages/utils/src/objects.ts#L67-L76

See previous discussion for more context https://github.com/ChainSafe/lodestar/pull/6749#discussion_r1609906528

**Comment by tinniaru3005** [comment]
Hey @nflaig, can i pick up this issue and work on it?

**Comment by csehatt741** [comment]
@nflaig, does the PR solve this issue? If further investigation is required just assign it to me!

**Comment by mkermani144** [comment]
@nflaig
It seems unassigned, can I pick it up?

**Comment by nflaig** [comment]
> [@nflaig](https://github.com/nflaig) It seems unassigned, can I pick it up?

yes, go ahead, thanks for picking it up, will assign you

**Comment by mkermani144** [comment]
Hey @nflaig 

Just checked the `unstable` branch, and noticed the code in the issue description is gone. No `client.ts` exists anymore, and searching the repo for similar function names, types, etc. didn't result in anything relevant.  

By the way, I found a similar usage here: 
 https://github.com/ChainSafe/lodestar/blob/011a0c4b89250b8fdc56a86f9b0a4698521554d0/packages/api/src/utils/client/method.ts#L49

Removing the type assertion in the above code just works. Do you know any other place that this issue raises typing errors without type assertion?

**Comment by nflaig** [comment]
I don't think removing the type assertions works, see the error

![Image](https://github.com/user-attachments/assets/28a6b86c-b8d4-4834-be71-a38c5ef38ef2)

the code is here https://github.com/ChainSafe/lodestar/blob/011a0c4b89250b8fdc56a86f9b0a4698521554d0/packages/api/src/utils/client/method.ts#L49

**Comment by nflaig** [comment]
> Removing the type assertion in the above code just works

how did you test it? maybe your local editor uses another typescript version which fixes the issue

> Do you know any other place that this issue raises typing errors without type assertion?

there shouldn't be but you can just check where the function is called

**Comment by mkermani144** [comment]
@nflaig Sorry for wrong data, I didn't build the project and that was the reason I didn't get an error.  

The error essentially is because we are trying to assign subtypes of a generic type constraint to the generic type parameter itself (here, via assigning `mapValues` returned value to `ApiClientMethods`). While you probably know it already, I'm not sure this is "fixable" via changing `mapValues` types or implementation. It's part of the language and how it works.  

The inlining works because it changes how Typescript infers types. But when wrapped, it changes how TS infers the types, and hence the error.

**Comment by nflaig** [comment]
> The inlining works because it changes how Typescript infers types. But when wrapped, it changes how TS infers the types, and hence the error.

yes, some type inference issue, not sure if it's fixable or not, part of this issue is to investigate, and if it turns out not fixable we can go ahead and close this, I personally don't think this type cast is a big deal

**Comment by mkermani144** [comment]
I think it's not easy to fix it without affecting references of `mapValues`, but someone may suggest a trick or something to bypass it. It makes sense to unassign it by the way, as I have no solution.

**Comment by nflaig** [comment]
> I think it's not easy to fix it without affecting references of `mapValues`, but someone may suggest a trick or something to bypass it. It makes sense to unassign it by the way, as I have no solution.

I unassigned you, thanks for taking a look

**Comment by nflaig** [comment]
well since the issue just says "investigate" I consider this closed, investigation has found that it's not solvable due to TS limitations
